torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1820,6 +1831,896 @@ def clip_grad_norm_(
+                if _is_zero_dim_tensor(step):
+                    param_state["step"] = step.item()
+        # Currently, we do not know of any other zero-dimension tensor state,
+        # so here, we check for such cases and error since it is unsupported;
+        # we can remove this if we choose to support zero-dimension tensors
+        for param_state in optim_state_dict["state"].values():
+            for state_name, state_value in param_state.items():
+                if state_name == "step":
+                    continue
+                if _is_zero_dim_tensor(state_value):
+                    raise ValueError(
+                        "Unsupported zero-dimension tensor state: "
+                        f"{state_name}"
+                    )
+        keys = set(optim_state_dict.keys())
+        assert keys == {"state", "param_groups"}, \
+            "Expects \"state\" and \"param_groups\" to be the keys in the " \
+            f"optimizer state dict but got {keys}"
+        return optim_state_dict
+
+    @staticmethod
+    def _unflatten_optim_state(
+        flat_param: FlatParameter,
+        flat_param_state: Dict[str, Any],
+        fsdp_module: "FullyShardedDataParallel",
+        target_rank: int,
+    ) -> Dict[int, Dict[str, Any]]:
+        """
+        Args:
+            flat_param (FlatParameter): Flattened parameter.
+            flat_param_state (Dict[str, Any]): Entry for a flattened parameter
+                in the "state" part of the optimizer state dict.
+            fsdp_module (FullyShardedDataParallel): FSDP module that owns the
+                flattened parameter, i.e. holds it in ``self.params``.
+
+        Returns:
+            unflat_param_state (Dict[int, Dict[str, Any]]): A :class:`dict`
+                holding the entries in the "state" part of the optimizer state
+                dict corresponding to the unflattened parameters comprising the
+                flattened parameter ``flat_param``. The keys are indexed
+                starting from zero and need to all be increased by an offset
+                for the overall optimizer state dict.
+        """
+        assert sum(p is flat_param for p in fsdp_module.params) == 1, \
+            "`fsdp_module` must own `flat_param`"
+        tensor_state, non_tensor_state, zero_dim_tensor_state, pad_per_rank = \
+            fsdp_module._communicate_optim_state(
+                flat_param, flat_param_state, target_rank,
+            )
+        unflat_param_state = fsdp_module._unflatten_communicated_optim_state(
+            flat_param,
+            tensor_state,
+            non_tensor_state,
+            zero_dim_tensor_state,
+            pad_per_rank,
+        )
+        return unflat_param_state
+
+    def _communicate_optim_state(
+        self,
+        flat_param: FlatParameter,
+        flat_param_state: Dict[str, Any],
+        target_rank: int = 0,
+    ) -> Tuple[Dict[str, List], Dict[str, List], Dict[str, List], List[int]]:
```

Comment:
Might be easier to read if this return object was a dataclass?

Also, I do think FSDP is getting a bit bloated with a lot of checkpointing related code. Could we possibly create a checkpoint_utils folder in fsdp, and add all of this logic (which isn't as pertinent to core FSDP execution code) to that to better separate things?